contracts/near/src/lib.rs

```diff
@@ -0,0 +1,437 @@
+            Promise::new(account_id).transfer(amount);
+        }
+    }
+
+    #[private]
+    pub fn set_upgrade_hash(&mut self, codehash: [u8; 32]) {
+        self.codehash = codehash;
+    }
+
+    /// Upon succesful verification of a VAA, this callback processes the VAA itself and branches
+    /// to the appropriate handler.
+    #[private]
+    pub fn process_vaa_callback(
+        &mut self,
+        account_id: AccountId,
+        vaa: String,
+        #[callback_result] _result: Result<u32, near_sdk::PromiseError>,
+    ) {
+        // Get Storage Usage before executed.
+        let storage_usage = env::storage_usage();
+
+        // Deserialize VAA, note that we already deserialized and verified the VAA in `process_vaa`
+        // at this point so we only care about the `rest` component which contains bytes we can
+        // deserialize into an Action.
+        let vaa = hex::decode(&vaa).unwrap();
+        let (_, rest): (structs::Vaa<()>, _) = vaa_payload::from_slice_with_payload(&vaa).unwrap();
+
+        match Action::try_from_slice(rest).unwrap() {
+            Action::AddSource(source) => self.add_source(source),
+            Action::DelSource(source) => self.del_source(source),
+            Action::UpgradeContract(codehash) => self.set_upgrade_hash(codehash),
+            Action::BatchAttest(batch) => {
+                let bytes = &mut Cursor::new(batch);
+                let batch = BatchPriceAttestation::deserialize(bytes).unwrap();
+                let count = self.attest_prices(&batch);
+                log!("Updated {} Price Feeds", count);
+            }
+        }
+
+        // Refund Storage Difference to `account_id` after storage execution.
+        let storage_diff = env::storage_usage() - storage_usage;
+        Promise::new(account_id).transfer(storage_diff as u128 * env::storage_byte_cost());
+    }
+
+    /// This method allows self-upgrading the contract to a new implementation. This function is
+    /// open to call by anyone, but to perform an authorized upgrade a VAA containing the hash of
+    /// the `new_code` must have previously been relayed to this contract's `process_vaa` endpoint.
+    /// otherwise the upgrade will fail.
+    ///
+    /// NOTE: This function is not pub so that it can only be called by the `upgrade_contract`
+    /// method, this is much much cheaper than serializing a Vec<u8> to call this method as a
+    /// normal public method.
+    #[handle_result]
+    fn upgrade(&mut self, new_code: Vec<u8>) -> Result<Promise, Error> {
+        let signature = env::sha256(&new_code);
+
+        if signature != self.codehash {
+            return Err(Error::UnauthorizedUpgrade);
+        }
+
+        Ok(Promise::new(env::current_account_id())
+            .deploy_contract(new_code)
+            .then(Self::ext(env::current_account_id()).refund_upgrade(
+                env::predecessor_account_id(),
+                env::attached_deposit(),
+                env::storage_usage(),
```

Comment:
When do the arguments to the callback get executed? Does this happen before or after the new contract is deployed?